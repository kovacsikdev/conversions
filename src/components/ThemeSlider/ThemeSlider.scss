.ThemeSlider {
  display: flex;
  align-items: center;
  margin: 4px;

  .switch {
    --knob-size: 22px;
    --slider-bg__light: #4d4d4d;
    --slider-bg__dark: #ccc;

    position: relative;
    width: 52px;
    height: 30px;

    img {
      width: 18px;
      margin: 6px;
      z-index: 10;
      position: absolute;
      pointer-events: none;
      left: 0;
      top: 0;
      bottom: 0;

      &:nth-child(1) {
        left: var(--knob-size);
      }
    }

    input {
      opacity: 0;
      width: 0;
      height: 0;

      &:checked + .slider {
        background-color: var(--slider-bg__dark);

        &:before {
          -webkit-transform: translateX(var(--knob-size));
          -ms-transform: translateX(var(--knob-size));
          transform: translateX(var(--knob-size));
        }
      }
    }

    .slider {
      position: absolute;
      cursor: pointer;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background-color: var(--slider-bg__light);
      -webkit-transition: 0.4s;
      transition: 0.4s;

      &:before {
        position: absolute;
        content: "";
        height: var(--knob-size);
        width: var(--knob-size);
        left: 4px;
        bottom: 4px;
        background-color: white;
        -webkit-transition: 0.4s;
        transition: 0.4s;
      }
    }

    .slider.round {
      border-radius: 34px;

      &:before {
        border-radius: 50%;
        z-index: 20;
      }
    }
  }
}
