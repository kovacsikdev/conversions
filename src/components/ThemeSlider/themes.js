export const themes = {
  light: {
    color: '#0d0d19',
    background: '#fff'
  },
  dark: {
    color: '#f2f2f2',
    background: '#0d0d19'
  },
  themeName: 'light'
};