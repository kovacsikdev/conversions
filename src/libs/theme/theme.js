/**
 * Object containing methods to either set or get data-theme from the document
 * @Object
 */
export const theme = {
  setTheme: (themeName) => {
    document.getElementsByTagName('html')[0].setAttribute('data-theme', themeName);
  },
  getTheme: () => document.getElementsByTagName('html')[0].getAttribute('data-theme')
}