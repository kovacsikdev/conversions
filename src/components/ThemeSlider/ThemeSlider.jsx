import React from 'react';
import { theme } from '../../libs/theme/theme'
import sunIcon from '../../assets/sun-icon.png';
import nightIcon from '../../assets/night-icon.png';
import './ThemeSlider.scss';

export const ThemeSlider = () => {
  const [themeName, setThemeName] = React.useState(theme.getTheme());

  const toggleTheme = () => {
    const newTheme = themeName === 'light' ? 'dark' : 'light';
    setThemeName(newTheme);
    localStorage.setItem('themeName', newTheme);
    theme.setTheme(newTheme);
  };

  return (
    <div className="ThemeSlider">
      <label className="switch">
        <img src={sunIcon} alt="light" />
        <input
          checked={themeName === 'dark'}
          type="checkbox"
          onChange={toggleTheme}
        />
        <span className="slider round"></span>
        <img src={nightIcon} alt="dark"/>
      </label>
    </div>
  )
}
