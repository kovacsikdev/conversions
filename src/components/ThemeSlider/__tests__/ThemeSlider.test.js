import React from 'react';
import { shallow } from 'enzyme';
import { ThemeSlider } from '../ThemeSlider';
import { theme } from '../../../libs/theme/theme';

theme.getTheme = jest.fn();
theme.setTheme = jest.fn();

describe("components/ThemeSlider", () => {
  beforeAll(() => {
    theme.getTheme.mockReturnValue('light');
    global.Storage.prototype.setItem = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    const wrapper = shallow(<ThemeSlider />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the theme from light to dark', () => {
    const wrapper = shallow(<ThemeSlider />);
    
    wrapper.find('input').simulate('change');

    expect(global.Storage.prototype.setItem).toHaveBeenCalledTimes(1);
    expect(global.Storage.prototype.setItem).toHaveBeenCalledWith('themeName', 'dark');
    expect(theme.setTheme).toHaveBeenCalledTimes(1);
    expect(theme.setTheme).toHaveBeenCalledWith('dark');
  });

  it('should toggle the theme from dark to light', () => {
    theme.getTheme.mockReturnValueOnce('dark');
    const wrapper = shallow(<ThemeSlider />);
    
    wrapper.find('input').simulate('change');

    expect(global.Storage.prototype.setItem).toHaveBeenCalledTimes(1);
    expect(global.Storage.prototype.setItem).toHaveBeenCalledWith('themeName', 'light');
    expect(theme.setTheme).toHaveBeenCalledTimes(1);
    expect(theme.setTheme).toHaveBeenCalledWith('light');
  });
});