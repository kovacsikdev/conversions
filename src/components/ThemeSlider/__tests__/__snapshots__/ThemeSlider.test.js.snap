// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ThemeSlider should render the component correctly 1`] = `
<div
  className="ThemeSlider"
>
  <label
    className="switch"
  >
    <img
      alt="light"
      src={
        Object {
          "canInstrument": true,
          "createTransformer": [Function],
          "getCacheKey": [Function],
          "process": [Function],
        }
      }
    />
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="slider round"
    />
    <img
      alt="dark"
      src={
        Object {
          "canInstrument": true,
          "createTransformer": [Function],
          "getCacheKey": [Function],
          "process": [Function],
        }
      }
    />
  </label>
</div>
`;
